File: array/permute.py
# leetcode 46. permute

import copy

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        path = []
        used = []
        self.backtrack(nums, path, used, res)
        return res
    
    def backtrack(self, nums, path, used, res):
        if len(path) == len(nums):
            res.append(copy.deepcopy(path))
            return
        
        for i in range(len(nums)):
            if nums[i] in used:
                continue
            path.append(nums[i])
            used.append(nums[i])
            self.backtrack(nums, path, used, res)
            path.pop()
            used.pop()

File: array/permute.cpp
// leetcode 46. permute

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) 
    {
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        this->backtrack(nums, path, used, res);
        return res;
    }

    void backtrack(vector<int> &nums, vector<int> &path, vector<bool> &used, vector<vector<int>> &res)
    {
        if (path.size() == nums.size())
        {
            res.push_back(path);
            return;
        }

        for (int i = 0; i < nums.size(); i++)
        {
            if (used[i])
            {
                continue;
            }
            path.push_back(nums[i]);
            used[i] = true;
            this->backtrack(nums, path, used, res);
            path.pop_back();
            used[i] = false;
        }
    }
};
